File: src/lib/spanishNumbers.ts
// Utilities to normalize and parse Spanish number phrases (0..400)

export function normalizeSpanish(text: string) {
    return text
        .toLowerCase()
        .replace(/á/g, 'a')
        .replace(/é/g, 'e')
        .replace(/í/g, 'i')
        .replace(/ó/g, 'o')
        .replace(/ú/g, 'u')
        .trim()
}

export function parseSpanishNumber(text: string): number | null {
    const t = normalizeSpanish(text)

    const units: Record<string, number> = {
        cero: 0,
        uno: 1,
        un: 1,
        una: 1,
        dos: 2,
        tres: 3,
        cuatro: 4,
        cinco: 5,
        seis: 6,
        siete: 7,
        ocho: 8,
        nueve: 9,
    }
    const teens: Record<string, number> = {
        diez: 10,
        once: 11,
        doce: 12,
        trece: 13,
        catorce: 14,
        quince: 15,
        dieciseis: 16,
        diecisiete: 17,
        dieciocho: 18,
        diecinueve: 19,
    }
    const tens: Record<string, number> = {
        veinte: 20,
        treinta: 30,
        cuarenta: 40,
        cincuenta: 50,
        sesenta: 60,
        setenta: 70,
        ochenta: 80,
        noventa: 90,
    }
    const hundreds: Record<string, number> = {
        cien: 100,
        ciento: 100,
        doscientos: 200,
        trescientos: 300,
        cuatrocientos: 400,
    }
    const fused20s: Record<string, number> = {
        veintiuno: 21,
        veintiun: 21,
        veintiuna: 21,
        veintidos: 22,
        veintitres: 23,
        veinticuatro: 24,
        veinticinco: 25,
        veintiseis: 26,
        veintisiete: 27,
        veintiocho: 28,
        veintinueve: 29,
    }

    if (t in fused20s) return fused20s[t]
    if (t in teens) return teens[t]
    if (t in units) return units[t]
    if (t in tens) return tens[t]
    if (t in hundreds) return hundreds[t]

    const tokens = t.split(/\s+/).filter(Boolean)
    const readUnit = (w?: string) => (w && w in units ? units[w] : 0)
    const readTens = (w?: string) => (w && w in tens ? tens[w] : 0)
    const readTeen = (w?: string) => (w && w in teens ? teens[w] : null)
    const readHundreds = (w?: string) => (w && w in hundreds ? hundreds[w] : 0)

    // hundreds + remainder
    const firstHundreds = readHundreds(tokens[0])
    if (firstHundreds) {
        const rest = tokens.slice(1)
        if (rest.length === 0) return firstHundreds
        const asTeen = readTeen(rest.join(' '))
        if (asTeen !== null) return firstHundreds + asTeen
        const tVal = readTens(rest[0])
        if (tVal) {
            if (rest.length >= 3 && rest[1] === 'y') return firstHundreds + tVal + readUnit(rest[2])
            return firstHundreds + tVal
        }
        const uVal = readUnit(rest[0])
        if (uVal) return firstHundreds + uVal
    }

    // tens + y + unit
    if (tokens.length >= 3 && tokens[1] === 'y' && tokens[0] in tens && tokens[2] in units) {
        return tens[tokens[0]] + units[tokens[2]]
    }
    // tens + unit without 'y'
    if (tokens.length === 2 && tokens[0] in tens && tokens[1] in units) {
        return tens[tokens[0]] + units[tokens[1]]
    }

    return null
}

File: scripts/spanish_numbers.py
"""Utilities for converting integers to their Spanish textual representation (0..300).

Rules implemented:
* 0 = "cero"
* 1..29 irregular / veinti* fused forms
* 30..99 => tens + " y " + unit (except exact tens)
* 100 exact => "cien"; 101..199 => "ciento <rest>"
* 200 => "doscientos"; 201..299 => "doscientos <rest>" (same for trescientos)
* 300 => "trescientos"

Scope purposely ends at 300 (no feminine agreement handling, no 400+).
"""
from __future__ import annotations

EXACT = {
    0: "cero",
    1: "uno",
    2: "dos",
    3: "tres",
    4: "cuatro",
    5: "cinco",
    6: "seis",
    7: "siete",
    8: "ocho",
    9: "nueve",
    10: "diez",
    11: "once",
    12: "doce",
    13: "trece",
    14: "catorce",
    15: "quince",
    16: "dieciséis",
    17: "diecisiete",
    18: "dieciocho",
    19: "diecinueve",
    20: "veinte",
    21: "veintiuno",
    22: "veintidós",
    23: "veintitrés",
    24: "veinticuatro",
    25: "veinticinco",
    26: "veintiséis",
    27: "veintisiete",
    28: "veintiocho",
    29: "veintinueve",
    30: "treinta",
    40: "cuarenta",
    50: "cincuenta",
    60: "sesenta",
    70: "setenta",
    80: "ochenta",
    90: "noventa",
    100: "cien",
    200: "doscientos",
    300: "trescientos",
}


def number_to_spanish(n: int, *, compound_style: str = "y") -> str:
    """Return Spanish cardinal for 0..300.

    compound_style:
        "y" -> use canonical form (e.g. "treinta y cinco")
        "split" -> return a tuple-like spaced form where user may want to fetch
                    separate audio segments (e.g. "treinta y cinco" unchanged
                    for now, but reserved if later we want segmentation logic).
    """
    if not (0 <= n <= 300):
        raise ValueError("Only numbers 0..300 supported")
    if n in EXACT:
        return EXACT[n]

    # 101..199 => ciento + remainder
    if 101 <= n <= 199:
        remainder = n - 100
        return f"ciento {number_to_spanish(remainder)}"

    # 201..299 => doscientos + remainder; 301 not allowed (cutoff 300)
    if 201 <= n <= 299:
        remainder = n - 200
        return f"doscientos {number_to_spanish(remainder)}"

    # 301 not supported; but handle 300 exact earlier

    # 30..99 & 1..99 general case if not exact already covered
    if n < 100:
        tens = (n // 10) * 10
        units = n % 10
        tens_word = EXACT[tens]
        return f"{tens_word} y {EXACT[units]}"

    raise AssertionError("Unreachable path for n<=300")


if __name__ == "__main__":  # Simple manual test
    test_points = [0, 1, 16, 21, 30, 31, 36, 99, 100, 101, 115, 121, 175, 199, 200, 201, 221, 275, 299, 300]
    for i in test_points:
        print(i, number_to_spanish(i))
